Core > Security: 블로렛 JWT 인가 필터 구현

wch-os opened the pull request:
## Issues

- Resolves #238

## Description

- 블로렛 서비스 JWT 인가 필터 구현
- 커스텀 어노테이션 + Argument Resolver 방식으로 인가 유저의 정보 관리
  - `@AuthUser`: 클라이언트 요청에 포함된 jwt 인가 정보를 AuthorizedUser 객체로 변환시켜주는 트리거 역할
  - `AuthorizedUser`: jwt 인가 정보를 담고 있는 객체

<br />

## Review Points

- `@AuthUser`와 `AuthorizedUser` 관련해서 좋은 네이밍이 생각나시면 추천 부탁드립니다!

<br />

## How Has This Been Tested?

- 테스트는 추가로 올리겠습니다..!! (테스트 작성이 어렵네요😭)

<br />

## Additional Notes

- 필터 + 리졸버 / 어노테이션 + 인가 객체 모듈 분리를 고민했습니다. 현재는 모놀리스라서 각 도메인에서 웹 요청과 같은 설정 정보를 관리하지 않아 의미가 있을 수 있겠지만, 이후 MSA와 같이 각 도메인에서 설정 정보를 관리해야 하므로 하나의 모듈로 관리하는 것이 좋을 것 같아 이와 같이 설계했습니다.

wch-os committed:
feat(core): add `blolet-jwt-filter` module

wch-os committed:
feat(blolet-jwt-fliter): add `AuthorizedUser` and `@AuthUser` for JWT injection

wch-os committed:
feat(blolet-jwt-filter): implement JWT authorization filter

wch-os committed:
feat(blolet-jwt-filter): add `@AuthUser` argument resolver

wch-os committed:
feat(monolith): add JWT based argument resolver

- JWT 토큰에서 인증된 사용자 정보를 추출하는 Argument Resolver를 추가

wch-os committed:
chore(blolet-jwt-filter): add README.md

wch-os committed:
refactor(blolet-jwt-filter): set JWT claims as request attributes instead of `AuthorizedUser`

merge-simpson commented on core/security/jwt/blolet-jwt-filter/src/main/java/nettee/jwt/filter/JwtFilterConfig.java:
```diff
@@ -0,0 +1,23 @@
+        "/auth/email/verification/check",
+        "/auth/token/refresh",
+        "/swagger-ui"
+    );
```
### 💊 Configuration Properties 사용을 검토

properties/yaml 파일에서 목록을 관리하는 방식도 고려하면 좋을 것 같아요.

<br />

**참고: 상수 제공**

대표적인 상수 제공 방식은 다음과 같습니다.

- `static final`: 하드 코딩 위주
- `enum`: (논외)
- Configuration properties: 소스 코드와 독립적인 관리 가능
  - 속성 파일을 통해 값 제공
  - 빈으로 제공
  - 프로파일마다 값을 나누어도 소스 코드에 영향 X

<br />

<table>
<tr>
  <th>방식</th>
  <th>특징</th>
</tr>
<tr>
  <td><strong>Static final</strong></td>
  <td>
  
  - 기본적으로 다른 빈에서 얻어 올 수 있는 정보를 사용할 수 없음.  
    : 안티패턴으로 억지로 주입은 가능하나 비권장
  - 기본적으로 빈으로 제공되지 않음.
  
  </td>
</tr>
<tr>
  <td><strong>Enum 열거 상수</strong></td>
  <td>
  
  - `static final`과 대체로 유사하게 활용 가능
  - 독립된 타입 선언으로 더 명시적인 선택지를 둔다는 차이점
  
  <details><summary>컴파일타임 보장</summary>
  <p>
  
  - 열거상수는 컴파일타임에 값 보장  
    (예를 들어 애노테이션 속성 값으로 넣을 수 있음.)
  - 열거상수의 내부 필드는 컴파일타임에 문법적으로는 값 보장 X  
    (논리적으로만 보장되는 상황이 많을 뿐, 예를 들어 애노테이션 속성 값으로 넣을 수 없음.)
    <h2 />
  
  </p>
  </details>
  
  <details><summary>상수별 지연 로딩 불가</summary>
  <p>
  
  - 한 enum 내 모든 열거상수가 동시에 메모리 적재
    
  <h2 />
  
  </p>
  </details> 
  
  </td>
</tr>
<tr>
  <td><strong>Configuration Properties</strong></td>
  <td>
  
  - 속성 파일을 통해 값 제공
  - 빈으로 제공
  - 프로파일마다 값을 나누어도 소스 코드에 영향 X
  
  </td>
</tr>
</table>

wch-os replied:
좋은 제안 감사합니다!

충분히 변경될 여지가 있는 path이기 때문에, properties 방향으로 수정하였습니다!

merge-simpson commented on core/security/jwt/blolet-jwt-filter/src/main/java/nettee/jwt/filter/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,111 @@
+            var claims = jwtParser.parseClaims(jwtToken);
+
+            // 4. JWT 토큰에서 사용자 정보 추출
+            request.setAttribute("userId", claims.getSubject());
+            request.setAttribute("roles", claims.get("roles", List.class));
+            request.setAttribute("profileIds", claims.get("profileIds", List.class));
```
<br />

## 💊 혹시 여기서 한 번에 객체로 저장하는 건 어떨까요?

> 지금 적용하는 방식은 사실 unsafe한 casting일 수 있습니다.

### 🚧 지금 방식의 주의 사항

1. Unchecked casting이므로 작업자의 주의가 필요합니다.
2. 다른 단계에서 request에 `userId`, `roles`, `profileIds`라는 이름으로 다른 변수를 담지 않아야 논리적으로 보장되는 캐스팅이므로 이름 중복이 없도록 주의해야 하죠.

한편 하나의 객체를 미리 만들어서 담으면 우려가 한층 줄어듭니다!
 
### ✨ Simple example

```java
// 👍 A: 특정 이름을 사용해도 기존 방식보다 전역적으로 더 고유한 역할을 표현하므로 충돌 확률이 낮음.
request.setAttribute("authorizedUser", authorizedUser);
```

### 💊 Better example

```java
// 👍 B: 조금 더 '안전한 캐스팅'을 유도하기 위해 클래스 풀네임을 문자열 키로 사용
//  K: "nettee.jwt.annotation.AuthorizedUser" (FQN: Full Qualified Name)
//  V: authoziredUser = AuthorizedUser.builder(). ... .build();
request.setAttribute(AuthorizedUser.class.getName(), authorizedUser);
```

이 예시들 또한 여전히 'unchecked casting'이지만, 여러 작업자가 작업을 보탤 때 생기는 실수를 많이 방지할 수 있습니다.

wch-os replied:
사실 두 가지 방법에 대해서 고민을 했습니다..😭

관련 커밋: 43b28cef44a539423e05ba14f2fab7361c9c6c50

- 고민을 하고 지은 결론은, "Filter에서는 순수하게 필드들을 담고, 이를 resolver에서 원하는 인가 객체(AuthorizedUser)에 맞게 생성해서 반환하자!" 였습니다. (Filter에서 인가 객체로의 의존을 최소화하기 위함)

---

> 다른 단계에서 request에 userId, roles, profileIds라는 이름으로 다른 변수를 담지 않아야 논리적으로 보장되는 캐스팅이므로 이름 중복이 없도록 주의해야 하죠.

- !!!!!! 감사합니다! 다른 필터에서 동일한 이름(userId, roles, profileIds)의 필드를 담을 가능성은 미처 생각하지 못했는데, 말씀해 주신 것처럼 중복에 유의해야 할 것 같습니다.
다만, 다른 필터에서 해당 필드를 사용해야 하는 경우 대안을 고민해보고 수정하는 것도 좋을 것 같습니다..!!

- 추가로 말씀주신 것처럼 두 가지 방법 모두 unchecked casting 이 발생하여, 주의를 한다는 전제 하에 구현 개발자에 대한 실수(?)는 크게 고려하지 않았습니다..!!

merge-simpson commented on core/security/jwt/blolet-jwt-filter/README.md:
````diff
@@ -0,0 +1,58 @@
+```json
+{
+  "error": "JWT 토큰이 만료되었습니다.",
+  "message": "Token has expired",
````
별건 아니지만, '액세스 토큰'이 역할을 조금 더 잘 표현하고 포괄적인 표현 같아요.

그래서 JWT 토큰보다 '액세스 토큰'이 어울린다고 느끼지만,
소통 측면에서는 사소한 차이라서 편하신 방향으로 해 주셔도 좋을 것 같습니다. 👍

merge-simpson commented on core/security/jwt/blolet-jwt-filter/src/main/java/nettee/jwt/filter/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,111 @@
+        String requestURI = request.getRequestURI();
+
+        return filterConfig.getExcludePaths().stream()
+                .anyMatch(requestURI::startsWith);
```
대안을 몇 가지 고려해 볼 수 있어 보여요.

**완전 일치만 비교 시**

- 컬렉션에서 바로 `contains` 사용  
  ```java
  return excludedPaths.contains(requestUri);
  ```  
  > 불필요한 스트림 인스턴스를 생성하지 않고, 포함 여부만 빠르게 확인.
- 이때 `excludedPaths`를 `HashSet` 타입 등으로 고려할 수도 있습니다.

**패턴 비교를 고려할 수도 있습니다.** (`startsWith` 대신)

- 패턴 비교가 느린 기미가 보이면, 추후 캐싱 라이브러리로 짧은 시간 캐싱해 둘 수 있습니다.  
  초기에는 `excludedPaths`의 사이즈가 크지 않으므로 별로 중요하지 않습니다.

wch-os replied:
좋은 제안 감사합니다! 

기존 'startsWith(접두사 일치)'일 때의 조건 판단을 진행했는데, 'contains(완전 일치)' 방식과 함께, List O(n) 복잡도에서, Set O(1) 복잡도로 수정하였습니다!

merge-simpson commented on monolith/main-runner/src/main/java/nettee/main/config/WebConfig.java:
MSA로 전환 시 이 파일을 중복으로 작성해야 하는 구조가 되므로, 중복을 피하기 위해 이 파일이 core 모듈 중 한 곳에 있으면 좋을 것 같습니다.

필요 시, configuration properties로 resolver 등록을 비활성화하도록 제어할 수도 있습니다.  
(지금은 불필요하고, 나중에 그러한 수정도 가능하다는 말입니다.)

wch-os replied:
확실히 다른 MSA로 전환되는 도메인에서 WebConfig가 중복으로 사용될 수 있겠군요!

해당 사항은 core > web-config 모듈 관련 이슈를 만들어서 반영하겠습니다!

merge-simpson commented on core/security/jwt/blolet-jwt-filter/src/main/java/nettee/jwt/resolver/AuthUserArgumentResolver.java:
```diff
@@ -0,0 +1,43 @@
+
+        String userId = (String) request.getAttribute("userId");
+        List<String> roles = (List<String>) request.getAttribute("roles");
+        List<String> profileIds = (List<String>) request.getAttribute("profileIds");
```
Unchecked casting이 반복되며, 사용된 명칭이 다른 API 엔드포인트에서 다루는 것과 추후 겹칠 가능성이 있어, 앞서 다른 코멘트로 언급드린 작업이 필요해 보입니다. (하나의 객체만 등록하는 등의 대안)

merge-simpson commented on core/security/jwt/blolet-jwt-filter/src/main/java/nettee/jwt/annotation/AuthorizedUser.java:
```diff
@@ -0,0 +1,37 @@
+ * @param roles      사용자 권한 목록
+ * @param profileIds 사용자 프로필 ID 목록
+ */
+public record AuthorizedUser(
+        String userId,
+        List<String> roles,
+        List<String> profileIds) {
```
콤팩트 생성자를 추가하여 `null` 대신 빈 문자열, 빈 리스트로 초기화하는 것은 어떨까요?
(참고: 이펙티브 자바 등)

**콤팩트 생성자의 일반적인 역할**

- 콤팩트 생성자답게 유효성 체크를 추가할 수도 있습니다.
- 또한 `null` 대신 빈 문자열과 빈 리스트를 사용함으로써 NPE(Null Pointer Exception) 일부를 피할 수 있습니다.

```suggestion
public record AuthorizedUser(
        String userId,
        List<String> roles,
        List<String> profileIds
) {
    // compact constructor (콤팩트 생성자)
    public AuthorizedUser {
        // validate parameters or set default value
        if (userId == null) {
            userId = "";
        }
        if (roles == null) {
            // if you want immutable list
            roles = List.of();
        }
        if (profileIds == null) {
            // if you want immutable list
            profileIds = List.of();
        }
    }
```

<br />

### 부연 설명: 콤팩트 생성자

*참고: 레코드의 전체 인자 생성자를 '캐노니컬 생성자'라고 합니다.

1. 모든 `record` 인스턴스는 반드시 전체 인자 생성자(all arguments constructor)에서 생성됩니다.  
  따라서 다른 생성자가 있으면 `this(...)` 생성자가 반드시 호출되어야 합니다.
    ```java
    public record Person(String name, boolean isAdult) {
        public Person(String name, int age) {
            // 전체 인자 생성자로 넘김
            this(name, age >= 18);
        }
    }
    ```
2. 콤팩트 생성자는 마치 캐노니컬 생성자의 진입점과 같습니다.  
    : 더 엄밀히 말하면, 콤팩트 생성자가 캐노니컬 생성자의 사전 동작만 작성하는 '콤팩트한' 생성자입니다.  
    (캐노니컬 생성자의 축약 버전으로, 컴파일러가 결국 캐노니컬로 확장하기 때문에 둘 중 하나만 작성 가능)
    ```java
    public record Person(String name, boolean isAdult) {
        // [2] 콤팩트 생성자
        public Person {
            // 파라미터의 검증 또는 전처리
            name = Objects.requireNonNull(name, "name must not be null.").strip();
            
            // 컴파일러가 자동으로 추가하는 동작:
            /*
            this.name = name;
            this.isAdult = isAdult;
            */
        }
        
        // [1] 다른 생성자들
        public Person(String name, int age) {
            // 전체 인자 생성자로 넘김
            this(name, age >= 18);
        }
    }
    ```

wch-os replied:
compact 생성자에서의 초기화를 생각치 못했네요!
감사합니다!

merge-simpson commented on core/security/jwt/blolet-jwt-filter/src/main/java/nettee/jwt/annotation/AuthorizedUser.java:
```diff
@@ -0,0 +1,37 @@
+        if (roles == null || roles.isEmpty()) {
+            return false;
+        }
+        return roles.contains(role);
```
### `isEmpty()`를 추가할 이유가 거의 없습니다.

`contains`가 내부적으로 `isEmpty()` 얼리리턴과 크게 다르지 않은 동작을 수행할 가능성이 큽니다.
(빈 컬렉션은 0회 순회)

따라서 다음과 같이 간략하게 작성할 수 있습니다. (성능도 비슷)

```suggestion
// 콤팩트 생성자 등에서 roles에 기본값 대입을 완료했다면 && 앞 조건도 생략 가능
return roles != null && roles.contains(role);
```

### `roles` 초깃값을 부여했을 때는 `null` 체크도 제외할 수 있습니다.

다음 쓰임새에서 NPE 방어 동작을 보장합니다.

```suggestion
return roles.contains(role);
```

wch-os replied:
compact 생성자를 사용하면서 non-null을 보장하며, 말씀주신 것처럼 `.isEmpty()` 동작이 무의미하여 
`.contains()` 만을 사용하여 boolean을 반환하도록 수정했습니다!

merge-simpson reviewed (approved):
@wch-os 님, 굉장히 깔끔하게 설계된 작업을 단기간에 잘해 주신 것 같습니다!

몇몇 부분은 꼭 바꾸었으면 좋겠지만, 지금 작업으로도 높은 완성도를 갖고 있어, approve를 미리 눌러 두려고 합니다!

수정 제안드린 부분들을 잘 고려해서 작업 부탁드리겠습니다!

**반드시 바꾸었으면 좋을 부분**

- `startsWith` 대신 URI 패턴으로 비교하기 (초기 성능 걱정 X)  
  : 인가는 빨라야 하지만, 추후 캐싱 등 대안을 도입할 수 있습니다.
- **리졸버 등록 위치 모듈**: MSA로 전환하더라도 각 마이크로서비스마다 중복된 작업을 생략할 수 있도록, 일부 코드는 `main-runner` 대신 코어 모듈에서 제공하면 좋겠습니다.  
  > WebFlux와 WebMVC에서 코드가 다를까 봐 코어 모듈에서 제외하고 실행 모듈에 넣으셨다면, WebFlux용은 보통 새로운 코어 모듈로 나누는 편이고, 필요 시 공용 데이터 스펙용 API 모듈을 한 층 추가로 나눌 수 있으므로 지금은 WebMVC용으로만 설계하셔도 좋습니다.
- `request` 객체에 낱개 속성을 넣고 있는 것  
  : 중복가능성이 존재하고 안전한 캐스팅 보장성이 낮은 unchecked casting 반복 대신, 중복 가능성이 낮은 key 사용, 안전한 캐스팅을 많이 유도하는 방식을 제안했습니다.

<br />

<table border="5" align="center">
<tr>
  <td>🍀</td>
  <td>⛳ Approve!</td>
</tr>
</table>

merge-simpson commented:
Status label에 `on-hold`를 발견했지만, 미리 리뷰를 남겼습니다!

혹시 리뷰를 받지 않는다고 명시하려면 draft로 전환 버튼을 눌러 둘 수 있습니다.

wch-os committed:
chore(blolet-jwt-filter): update README.md

wch-os committed:
feat(blolet-jwt-filter): optimize JWT filter path exclusion

Changes the `excludePaths` from a List to a Set to ensure O(1) complexity for path lookups.

wch-os committed:
feat(blolet-jwt-filter): ensure non-null fields in `AuthorizedUser`

wch-os committed:
chore(blolet-jwt-filter): Renames AuthenticatedUser to AuthorizedUser

wch-os committed:
feat(blolet-jwt-filter): configure JWT filter using properties

wch-os committed:
feat(blolet-jwt-filter): add @SuppressWarnings("unchecked") for unchecked cast warnings

wch-os committed:
Merge remote-tracking branch 'origin/develop' into feature/core-blolet-jwt-filter

merge-simpson reviewed (approved):
👍 LGTM

@wch-os 님, 바쁘신 날로 들었는데 정말 고생 많으셨습니다!
병합해 주셔도 좋을 것 같습니다.

추가로 떠오르는 것들은 새로운 이슈로 제안하겠습니다!
감사합니다!

wch-os committed:
refactor(blolet-jwt-filter): updates config prefix for exclude paths

wch-os committed:
refactor(jwt): reorganize packages and rename config class

wch-os committed:
feat(monolith): add local JWT config
